cranelift/wasm/src/code_translator.rs

```diff
@@ -1941,9 +1941,20 @@ fn translate_unreachable_operator<FE: FuncEnvironment + ?Sized>(
             let control_stack = &mut state.control_stack;
             let frame = control_stack.pop().unwrap();
 
-            // Now we have to split off the stack the values not used
-            // by unreachable code that hasn't been translated
-            stack.truncate(frame.original_stack_size());
+            // Pop unused parameters from stack.
+            //
+            // As above in `translate_operator()`, we need to account for the duplication of
+            // parameters.
+            let num_duplicated_params = if let ControlStackFrame::If {
+                num_param_values, ..
+            } = frame
+            {
+                debug_assert!(num_param_values <= frame.original_stack_size());
+                num_param_values
+            } else {
+                0
+            };
+            stack.truncate(frame.original_stack_size() - num_duplicated_params);
```

Comment:
Random drive by, since this is the duplication might have a small code smell: would it make sense to add an helper on the ControlStackFrame struct itself, so we could do `frame.truncate_to_original_size(&mut stack)`, and hide the details of the number of parameters? Then the `original_stack_size` helper would be redundant, if I remember correctly...

Reply:
Good call, thanks! Factored out into a helper.

Reply:
Good call indeed, much cleaner now!